### PR TITLE
Follow Path with googlemaps directions api integration (follows roads and paths)

### PR DESCRIPTION
**Pull request by LasseSkogland**
Short Description: 
Added a new cell_worker based on `pokemongo_bot/cell_workers/follow_path.py`
It utilizes the Google Maps Directions API to follow roads and paths.
This is a "proof of concept" and could probably use some more work.
I will run some test to see how stable the code is. (it should be stable)

Fixes:
- Added  Class FollowHumanPath

### CODE REVIEW

**Comment by solderzzc** [comment]
This is nice, in my option, the follow path and follow GWalking should be separated, how do you think about?

**Comment by LasseSkogland** [comment]
They are seperated.
FollowPath in `pokemongo_bot/cell_workers/follow_path.py`
FollowHumanPath in `pokemongo_bot/cell_workers/follow_path.py`

**Comment by solderzzc** [comment]
@LasseSkogland  Yes, I just figured it out.
Could you please have the config example updated?

**Comment by solderzzc** [comment]
Which kind of key should be used, the JS key?

File: pokecli.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 import ssl
 import sys
 import time
+import googlemaps
```
Comment (solderzzc) [inline pokecli.py]: We don't need import the googlemaps here, right?

File: pokemongo_bot/__init__.py [hunk 2]
```diff
@@ -34,6 +35,10 @@ def position(self):
 
     def __init__(self, config):
         self.config = config
+        try:
+            self.gmap_client = googlemaps.Client(config.gmapkey)
```
Comment (solderzzc) [inline pokemongo_bot/__init__.py]: The initial should better be inside Task.

**Comment by LasseSkogland** [comment]
JS Key? Do you mean userdata,js.
I get the api key from config.json

**Comment by solderzzc** [comment]
I mean the google map key type, it should be js key to be used inside the googlemaps package

File: configs/config.json.humanpath.example [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        "type": "CatchLuredPokemon"
+      },
+      {
+        "type": "SpinFort"
+      },
+      {
+        "type": "FollowHumanPath",
+        "config": {
+          "path_mode": "linear",
+          "path_file": "configs/path.example.json"
```
Comment (solderzzc) [inline configs/config.json.humanpath.example]: The google path waypoint has 21 points limitation, is that a problem?

**Comment by MZorzy** [comment]
There are too much player for niantic to check if bot is not on road, and it can be on park.
but maybe will be a check if player cross a big building (minus gps lag/errors) so follow map can be good.

**Comment by solderzzc** [comment]
@MZorzy  I think the GPX configuration is not user friendly, should we just calc the best route in a specific area instead of the GPX input? The bot will be much more AI driving.

**Comment by LasseSkogland** [comment]
The way I do it is getting the overview_polyline points, but it isn't the most accurate.
I'll work on using the more accurate polyline points under ['routes']['legs']['steps']

**Comment by solderzzc** [comment]
I have code since the project started, I intended to use GWalk for the walk. 
We can draw a route connected 'Spawn Point'.'Fort' one by one, and loop in it or jump to another cell when done according the configuration.
```
def pokemon_working():
    gmaps = googlemaps.Client(key=GOOGLEMAPS_KEY)
    directions_result = gmaps.directions((49.004, 8.456),
                                         (49.004, 8.469),
                                         mode="walking")
    if directions_result and len(directions_result) > 0:
        """
        print directions_result[0]
        for ka, va in directions_result[0].iteritems():
            print ka+'\n'
            print va
        exit(0)
        """
        steps = directions_result[0]['legs'][0]['steps']
        print len(steps)
        for index, item in enumerate(steps):
            print index,item['start_location'],'-->',item['end_location'],'duration: ',item['duration']['value'],'sec'
```

File: pokemongo_bot/__init__.py [hunk 4]
```diff
@@ -103,7 +103,7 @@ def get_meta_cell(self):
                 wild_pokemons += cell["wild_pokemons"]
             if "catchable_pokemons" in cell and len(cell["catchable_pokemons"]):
                 catchable_pokemons += cell["catchable_pokemons"]
-        
+
```
Comment (solderzzc) [inline pokemongo_bot/__init__.py]: This code should be clean up.

**Comment by LasseSkogland** [comment]
Preview: (marker is original position, lines are GMaps Directions API logic)
image

File: pokemongo_bot/cell_workers/follow_human_path.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    def load_path(self):
+        if self.path_file is None:
+            raise RuntimeError('You need to specify a path file (json or gpx)')
+
+        path = []
+        if self.path_file.endswith('.json'):
+            path = self.load_json()
+        elif self.path_file.endswith('.gpx'):
+            path = self.load_gpx()
+        if not self.gmap_client == None:
```
Comment (solderzzc) [inline pokemongo_bot/cell_workers/follow_human_path.py]: If there's no path pass in, we should better calc the path from the cell list.

**Comment by TheSavior** [comment]
I thought this is what the polyline walker is supposed to do. https://github.com/PokemonGoF/PokemonGo-Bot/blob/master/pokemongo_bot/polyline_walker/polyline_walker.py#L11

We already have this. We need our navigators to use the polyline walker instead of the stepwalker and we are golden. It probably just needs a config flag. @th3w4y Can you provide some more context on the polyline walker and its usage please?

File: pokemongo_bot/cell_workers/follow_human_path.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+from pokemongo_bot.cell_workers.base_task import BaseTask
+from pokemongo_bot.cell_workers.utils import distance, i2f, format_dist
+from pokemongo_bot.human_behaviour import sleep
+from pokemongo_bot.step_walker import StepWalker
+from pgoapi.utilities import f2i
+
+
+class FollowHumanPath(BaseTask):
+    def get_gmap_directions(self, origin, destination):
+        directions_json = googlemaps.directions.directions(self.gmap_client, origin, destination, mode="walking", optimize_waypoints=True)
```
Comment (th3w4y) [inline pokemongo_bot/cell_workers/follow_human_path.py]: optimize_waypoints is not needed since you don't give waypoints

**Pull request closed without merging**